atd-vzd/data_model_readme.md

```diff
@@ -0,0 +1,369 @@
+
+#### 9. Add a new editable column to crashes
+
+To accomplish this, we must add the column to all three crashes tables, and update three triggers (CRIS insert, CRIS update, VZ update).
+
+```sql
+-- create the new column x 3
+alter table cris_crashes add column is_txdot_awesome boolean default true;
+alter table vz_crashes add column is_txdot_awesome boolean;
+alter table crashes add column is_txdot_awesome boolean;
+
+-- patch function that runs on CRIS insert
+create or replace function db.cris_crash_insert_rows()
+returns trigger
+language plpgsql
+as
+$$
+BEGIN
+    RAISE NOTICE 'Inserting vz_crashes and crashes rows';
+    -- insert new (editable) vz record (only crash ID)
+    INSERT INTO db.vz_crashes (crash_id) values (new.crash_id);
+    -- insert new combined / official record
+    INSERT INTO db.crashes (crash_id, primary_address, road_type_id, latitude, longitude) values (
+        new.crash_id, new.primary_address, new.road_type_id, new.latitude, new.longitude, new.is_txdot_awesome
+    );
+    RETURN NULL;
+END;
+$$;
+
+-- patch function that runs on VZ crash update
+create or replace function db.vz_crash_update()
+returns trigger
+language plpgsql
+as $$
+BEGIN
+    RAISE NOTICE 'Refreshing crash ID % due to vz_crashes update', new.crash_id;
+    UPDATE
+        db.crashes
+    SET
+        primary_address = COALESCE(new.primary_address, cris_crashes.primary_address),
+        road_type_id = COALESCE(new.road_type_id, cris_crashes.road_type_id),
+        latitude = COALESCE(new.latitude, cris_crashes.latitude),
+        longitude = COALESCE(new.longitude, cris_crashes.longitude),
+        is_txdot_awesome = COALESCE(new.is_txdot_awesome, cris_crashes.is_txdot_awesome)
+    FROM (
+        SELECT
+            *
+        FROM
+            db.cris_crashes where cris_crashes.crash_id = new.crash_id) AS cris_crashes
+WHERE
+    crashes.crash_id = new.crash_id;
+    RETURN NULL;
+END;
+$$;
+
+
+-- patch function that runs on CRIS crash update
+create or replace function db.cris_crash_update()
+returns trigger
+language plpgsql
+as $$
+DECLARE
+   vz_record  record;
+BEGIN
+    RAISE NOTICE 'Updating crash ID % due to crish_crashes update', new.crash_id;
+
+    SELECT INTO vz_record *
+        FROM db.vz_crashes where crash_id = new.crash_id;
+
+    UPDATE
+        db.crashes
+    SET
+        primary_address = COALESCE(vz_record.primary_address, NEW.primary_address),
+        road_type_id = COALESCE(vz_record.road_type_id, NEW.road_type_id),
+        latitude = COALESCE(vz_record.latitude, NEW.latitude),
+        longitude = COALESCE(vz_record.longitude, NEW.longitude),
+        is_txdot_awesome = COALESCE(vz_record.is_txdot_awesome, NEW.is_txdot_awesome)
+    WHERE
+        crashes.crash_id = new.crash_id;
+    RETURN NULL;
+END;
+$$;
+```
+
+And now we can use the new column :) 
+
+```sql
+update vz_crashes set is_txdot_awesome = false where crash_id = 9999999;
+```
+ 
+#### 10. Conflict management
```

Comment:
I like the idea of moving the conflict logic into triggers in the DB since we could observe the conditions for inserts into the conflict table more easily imo (even if it gets long).